Sherlock/sherlock.h

```diff
@@ -124,30 +124,40 @@ SOFTWARE.
 
 namespace sherlock {
 
-// TODO(dkorolev): Make persistence strategy a template parameter and initialize it properly.
-using blocks::persistence::MemoryOnly;
-
-template <typename T>
+template <typename ENTRY, template <typename, typename> class PERSISTENCE_LAYER, class CLONER>
 class StreamInstanceImpl {
  public:
-  StreamInstanceImpl(const std::string& name,
-                     const std::string& value_name,
-                     const std::function<T(const T&)> clone)
-      : name_(name), value_name_(value_name), storage_(std::make_shared<MemoryOnly<T>>(clone)) {
-    // TODO(dkorolev): Register this stream under this name.
-    // TODO(dk+mz): Ensure the stream lives forever.
-  }
+  typedef ENTRY T_ENTRY;
+  typedef PERSISTENCE_LAYER<ENTRY, CLONER> T_PERSISTENCE_LAYER;
 
-  // `Publish()` and `Emplace()` return the index of the added entry.
-  size_t Publish(const T& entry) { return storage_->Publish(entry); }
+  // TODO(dkorolev): In constuctor: Register the stream under its name in a singleton.
+  // TODO(dkorolev): In constuctor: Ensure the stream lives forever.
+
+  StreamInstanceImpl(const std::string& name)
+      : name_(name), storage_(std::make_shared<T_PERSISTENCE_LAYER>()) {}
 
-  size_t Publish(T&& entry) { return storage_->Publish(std::move(entry)); }
+  template <typename EXTRA_PARAM>
+  StreamInstanceImpl(const std::string& name, EXTRA_PARAM&& extra_param)
+      : name_(name), storage_(std::make_shared<T_PERSISTENCE_LAYER>(std::forward<EXTRA_PARAM>(extra_param))) {}
 
-  template <typename E>
-  typename std::enable_if<bricks::can_be_stored_in_unique_ptr<T, E>::value, size_t>::type Publish(const E& e) {
-    // TODO(dkorolev): Don't rely on the existence of copy constructor.
-    // TODO(dkorolev): Eliminate this copy.
-    return storage_->Emplace(new E(e));
+  // `Publish()` and `Emplace()` return the index of the added entry.
+  // Deliberately keep these two signatures and not one with `std::forward<>` to ensure the type is right.
+  size_t Publish(const ENTRY& entry) { return storage_->Publish(entry); }
+  size_t Publish(ENTRY&& entry) { return storage_->Publish(std::move(entry)); }
+
+  // When `ENTRY` is n `unique_ptr<>`, support two more syntaxes for `Publish`-ing entries of derived types.
```

Comment:
:+1: rephrased in 3c53467

It was "is an `std::unique_ptr`", then I shortened the line by removing `std::`, and then erased the wrong letter from `an` :-(